node/pkg/p2p/ccq_p2p.go

```diff
@@ -0,0 +1,297 @@
+			return false
+		}))
+	if err != nil {
+		return fmt.Errorf("failed to create new gossip sub for req: %w", err)
+	}
+
+	// We want to join and subscribe to the request topic. We will receive messages from there, but never write to it.
+	ccq.th_req, err = ps.Join(topic_req)
+	if err != nil {
+		return fmt.Errorf("failed to join topic_req: %w", err)
+	}
+
+	// We only want to join the response topic. We will only write to it.
+	ccq.th_resp, err = ps.Join(topic_resp)
+	if err != nil {
+		return fmt.Errorf("failed to join topic_resp: %w", err)
+	}
+
+	// We only want to accept messages from peers in the allow list.
+	err = ps.RegisterTopicValidator(topic_req, func(ctx context.Context, from peer.ID, msg *pubsub.Message) bool {
+		if len(ccq.allowedPeers) == 0 {
+			return true
+		}
+		if _, found := ccq.allowedPeers[from.String()]; found {
+			return true
+		}
+		ccq.logger.Debug("Dropping message from unknown peer", zap.String("fromPeerID", from.String()))
+		return false
+	})
+	if err != nil {
+		return fmt.Errorf("failed to register message filter: %w", err)
+	}
+
+	// Increase the buffer size to prevent failed delivery to slower subscribers
+	ccq.sub, err = ccq.th_req.Subscribe(pubsub.WithBufferSize(1024))
+	if err != nil {
+		return fmt.Errorf("failed to subscribe topic_req: %w", err)
+	}
+
+	common.StartRunnable(ctx, errC, false, "ccqp2p_listener", func(ctx context.Context) error {
+		return ccq.listener(ctx, signedQueryReqC)
+	})
+
+	common.StartRunnable(ctx, errC, false, "ccqp2p_publisher", func(ctx context.Context) error {
+		return ccq.publisher(ctx, gk, queryResponseReadC)
+	})
+
+	ccq.logger.Info("Node has been started", zap.String("peer_id", ccq.h.ID().String()), zap.String("addrs", fmt.Sprintf("%v", ccq.h.Addrs())))
+	return nil
+}
+
+func (ccq *ccqP2p) close() {
+	ccq.logger.Info("entering close")
+
+	if err := ccq.th_req.Close(); err != nil && !errors.Is(err, context.Canceled) {
+		ccq.logger.Error("Error closing the topic_req", zap.Error(err))
+	}
+	if err := ccq.th_resp.Close(); err != nil && !errors.Is(err, context.Canceled) {
+		ccq.logger.Error("Error closing the topic_req", zap.Error(err))
+	}
+
+	ccq.sub.Cancel()
+
+	if err := ccq.h.Close(); err != nil {
+		ccq.logger.Error("error closing the host", zap.Error(err))
+	}
+}
+
+func (ccq *ccqP2p) listener(ctx context.Context, signedQueryReqC chan<- *gossipv1.SignedQueryRequest) error {
+	for {
+		envelope, err := ccq.sub.Next(ctx) // Note: sub.Next(ctx) will return an error once ctx is canceled
+		if err != nil {
+			return fmt.Errorf("failed to receive pubsub message: %w", err)
+		}
+
+		var msg gossipv1.GossipMessage
+		err = proto.Unmarshal(envelope.Data, &msg)
+		if err != nil {
+			ccq.logger.Info("received invalid message",
+				zap.Binary("data", envelope.Data),
+				zap.String("from", envelope.GetFrom().String()))
+			ccqP2pMessagesReceived.WithLabelValues("invalid").Inc()
+			continue
+		}
+
+		ccq.logger.Info("received message", //TODO: Change to Debug
```

Comment:
Is now the time to do this?

Reply:
No, would like to leave it info at least until we're ready to go to mainnet.